### DISCUSSION TITLE
Booting an ISO via sanboot or memdisk

### DISCUSSION THREAD

**Discussion by n3gwg** [ipxe/ipxe#962] [General]
iPXE users, developers, et alia:

If I wish to setup an environment (in a VM lets say) to host ISOs to be able to boot via iPXE. It became known to me that the easiest way (for testing at least) was to run "ruby -run -ehttpd . -p8080" from the directory with the ISOs in it. Indeed, iPXE was able to access the directory to try to load an ISO, but alas, I was unable to boot anything.

I tried several different combinations of sanboot and memdisk.

If anyone has any recommendations of what they got to work, I'd like to know so I can try it (I know not all ISOs are created the same, so there is no guarantee if it worked for you it will work for me).

I am booting using netboot.xyz.efi and it has the following:

Features: DNS HTTP HTTPS iSCSI TFTP VLAN SRP AoE EFI Menu

One more thing to note, I did try this code:
initrd http://1.2.3.4:8080/abc.iso || shell
chain http://1.2.3.4:8080/memdisk iso raw || shell

The error I got was:
Could not boot: Exec format error (https://ipxe.org/2e008081)

Thanks in advance!

Stuart

**Comment by stappersg** [comment]
FWIW, from the _iPXE IRC Channel_
```text
01:01  * stappers retrieves from "IRC command line history" ...
01:01 < stappers> <https://libera.ems.host/_matrix/media/v3/download/libera.chat/7aa4a8640b0f3fb6036b7fcc8a95d0c9af457b97>)
01:02 < stappers> guest96: For what it is worth: Booting from ISO and networkbooting are separate things,
    mixing them results in a mix. (And understanding a mix is indeed hard.)
01:03 < stappers> done
 [01:03] [stappers(+iw)] [5:libera/#ipxe(+nt)]                                                                          
```

**Reply by stappersg** [reply]
And text from https://ipxe.org/2e008081 :
> If you are running a UEFI version of iPXE then only EFI binaries can be loaded. PXE binaries such as pxelinux.0 and memdisk use legacy BIOS calls and can never work on UEFI.

**Reply by n3gwg** [reply]
I did not realize that memdisk was in fact a PXE binary. Does that mean that if I boot to PXE I could use memdisk to load the ISO? I imagine I'd need to create an iPXE entry that rebotted me over using PXE booting somehow?

I did a bit more reading (though am still not entirely clear yet), but it seems that if boot EFI then I can load an ISO via HTTP? If so, do you know where I can get code that will let me do that?

Stuart

**Comment by NiKiZe** [comment]
Memdisk is a PXE binary, not available in EFI.
what it does is to emulate CD access from data stored in RAM.
This means it downloads the full ISO, without maybe using the full contents.
The next thing is, that as soon as the OS boots, the references to the existence of "a CD" is totally gone. Which means that any boot most often fails at this stage. (Exception being special memdisk aware kernels)

The same goes for sanboot over http, with the difference that the ISO data is only downloaded as needed.
If using iSCSI, and booting iSCSI aware OS, then it might work.

TL;DR: don't boot ISOs, a quick search here on GitHub for sanboot will give you discussions and issues explaining this even further.

**Comment by n3gwg** [comment]
Just so I can understand you better, I am curious; there is a project netboot.xyz that uses iPXE and all it does is boot tons of Linux ISOs, and it works using EFI. How does it get the job done then?

Stuart

**Reply by stappersg** [reply]
> there is a project netboot.xyz that uses iPXE and all it does is boot tons of Linux ISOs, and it works using EFI

Come again

**Reply by NiKiZe** [reply]
They don't boot ISOs in EFI for the most part, but with that said, some ISOs is only used in firmware mode, not not needed after kernel boots, in that case it works.
